src/org/labkey/test/components/ui/navigation/apps/ServerNotificationMenu.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * Click the 'Mark all as read' link. This will cause the list to refresh, so return a new reference to the
+     * notification list.
+     *
+     * @return A new list of {@link ServerNotificationItem}.
+     */
+    public List<ServerNotificationItem> clickMarkAll()
+    {
+        elementCache().markAll().click();
+        return getNotifications();
+    }
+
+    /**
+     * Expand the menu and return the notification.
+     *
+     * @return A list of {@link ServerNotificationItem}, empty if there are none.
+     */
+    public List<ServerNotificationItem> getNotifications()
+    {
+        if(!isExpanded())
+            expand();
+
+        if(getWrapper().isElementPresent(elementCache().notificationsContainerLocator))
+        {
+            return elementCache().notifications();
+        }
+        else
+        {
+            return new ArrayList<>();
+        }
+
+    }
+
+    /**
+     * If no notifications get the message that is present.
+     *
+     * @return Message shown if there are no notifications, and empty string of there are notifications.
+     */
+    public String getNoNotificationsMessage()
+    {
+
+        if(!isExpanded())
+            expand();
+
+        if(elementCache().noNotificationsElement().isDisplayed())
+        {
+            return elementCache().noNotificationsElement().getText();
+        }
+        else
+        {
+            return "";
+        }
+
+    }
+
+    @Override
+    protected Locator getToggleLocator()
+    {
+        return Locator.tagWithId("button", "server-notifications-button");
+    }
+
+    @Override
+    protected ElementCache elementCache()
+    {
+        return (ElementCache) super.elementCache();
+    }
+
+    @Override
+    protected ElementCache newElementCache()
+    {
+        return new ElementCache();
+    }
+
+    protected class ElementCache extends BaseBootstrapMenu.ElementCache
+    {
+        private final Locator notificationsContainerLocator = Locator.tagWithClass("div", "server-notifications-listing-container");
+
+        public final WebElement notificationsContainer()
+        {
+            return notificationsContainerLocator.findWhenNeeded(this);
+        }
+
+        public final List<ServerNotificationItem> notifications()
+        {
+            return new ServerNotificationItem.AppNotificationEntryFinder(getDriver()).findAll(notificationsContainer());
+        }
+
+        public final WebElement statusIcon()
+        {
+            return Locator.tagWithClass("i", "fa").findWhenNeeded(this);
+        }
+
+        public final WebElement noNotificationsElement()
+        {
+            return Locator.tagWithClass("div", "server-notifications-footer").findWhenNeeded(this);
+        }
+
+        public final WebElement markAll()
+        {
+            return Locator.tagWithClass("h3", "navbar-menu-header")
+                    .child("div[contains(@class,'server-notifications-link')]").findWhenNeeded(this);
```

Comment:
Using locator builders will generate CSS Locators which are, generally, faster that XPath.
```suggestion
            return Locator.tagWithClass("h3", "navbar-menu-header")
                    .child(Locator.tagWithClass("div", "server-notifications-link").findWhenNeeded(this);
```

Reply:
Changed.
I tried to avoid the XPath but couldn't get the right locator. Thanks!